src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -119,78 +121,104 @@ SymbolKind.NetModule or
                     {
-                        // We may still have a cached generator; we'll have to remember to run generators again since we are making some
-                        // change here. We'll also need to update the other state of the driver if appropriate.
-
-                        // The no compilation state can never be in the 'DocumentsAreFinal' state.  The only place where
-                        // we start with the NoCompilationState is the 'Empty' instance (where DocumentsAreFinal=false).
-                        // And then this is the only place where we get a NoCompilationState and create a new instance.
-                        // So there is no way to ever transition this to the DocumentsAreFinal=true state.
-                        Contract.ThrowIfTrue(state.GeneratorInfo.DocumentsAreFinal);
-                        var generatorInfo = state.GeneratorInfo;
-                        if (generatorInfo.Driver != null && translate != null)
-                            generatorInfo = generatorInfo.WithDriver(translate.TransformGeneratorDriver(generatorInfo.Driver));
-
-                        return new NoCompilationState(generatorInfo);
+                        return ForkNoCompilationState(noCompilationState, translate);
                     }
                     else
                     {
                         throw ExceptionUtilities.UnexpectedValue(state.GetType());
                     }
                 }
+
+                static NonFinalWithCompilationTrackerState ForkWithCompilationTrackerState(
+                    ProjectState oldProjectState,
+                    WithCompilationTrackerState state,
+                    CompilationAndGeneratorDriverTranslationAction? translate)
+                {
+                    var finalCompilationWithGeneratedDocuments = state is FinalCompilationTrackerState finalState
+                        ? finalState.FinalCompilationWithGeneratedDocuments
+                        : null;
+
+                    var intermediateProjects = UpdateIntermediateProjects(oldProjectState, state, translate);
+
+                    var newState = CompilationTrackerState.Create(
+                        state.IsFrozen,
+                        state.CompilationWithoutGeneratedDocuments,
+                        state.GeneratorInfo,
+                        finalCompilationWithGeneratedDocuments,
+                        intermediateProjects);
+                    return newState;
+                }
+
+                static NoCompilationState ForkNoCompilationState(
+                    NoCompilationState state,
+                    CompilationAndGeneratorDriverTranslationAction? translate)
+                {
+
+                    // We may still have a cached generator; we'll have to remember to run generators again since we are making some
+                    // change here. We'll also need to update the other state of the driver if appropriate.
```

Comment:
Per other comments and conversations, can this happen anymore?

Reply:
no.  note: i have a followup pr that just cleans up NoCompilationState entirely.